src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php

```diff
@@ -394,7 +394,7 @@ public function relationsToArray()
             // If the relation value has been set, we will set it on this attributes
             // list for returning. If it was not arrayable or null, we'll not set
             // the value on the array because it is some type of invalid value.
-            if (isset($relation) || is_null($value)) {
+            if (isset($relation)) {
```

Comment:
Doesn't this change the behavior? `null` values should enter the condition. Removing the `|| is_null()` they will not enter anymore as the `isset()` without the `is_null()` check would exclude them. 

```php
$relation, $value = null;

if (isset($relation)) {
// we don't reach this code
}

if (isset($relation) || is_null($value)) {
// this code is executed
}

If this is some kind of bug fixing it shouldn't be part of this PR IMO.

Reply:
This yielded the following error:

```
 ------ ------------------------------------------------------------------------------------------------------------------ 
  Line   Illuminate/Database/Eloquent/Concerns/HasAttributes.php (in context of class Illuminate\Database\Eloquent\Model)  
 ------ ------------------------------------------------------------------------------------------------------------------ 
  398    Variable $relation might not be defined.                                                                          
         🪪  variable.undefined         
```

The `is_null($value)` is already checked above:

https://github.com/laravel/framework/blob/46297e441252d495713fb30223ffcfb325a4f120/src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php#L380-L385

And you'll notice that `$value` is not referenced inside the `if` body and therefore should not be referenced inside the condition. The only thing that matters is checking that the `$relation` isset before assigning it.

Reply:
Hey @calebdw I think we are talking different things here. 

The first check for `is_null($value)` just sets the `$relation` value to `null`. The check you removed sets `$attributes[$key] = $relation;` when `$value = null **OR** isset($relation)`. 

The fact that `$value` is not used inside the if, does not mean you can remove it from the condition as that would turn up not setting `$attributes[$key] = null` like was previously done.

Reply:
If `$value === null`, then `$relation` isset.

Reply:
Hmm, apparently there's some idiosyncrasies with [isset](https://www.php.net/manual/en/function.isset.php):

> Determine if a variable is considered set, this means if a variable is declared and is different than [null](https://www.php.net/manual/en/reserved.constants.php#constant.null).

Reply:
It turns out there's not a straightforward way in PHP to check if a variable isset / declared (null values being acceptable). Even using `isset($relation) || is_null($relation)` fails because `is_null` returns true when the variable does not exist :roll_eyes: .

Thanks for pointing this out! I also added a test since this is rather tricky

Reply:
Well, theoretically you can do a `get_defined_vars()` and a `array_key_exists()` to check if a variable is actually defined disregarding it's value: 

```php
$test = null;

isset($test); // return false
array_key_exists('test', get_defined_vars()); // return true
```

Note that `get_defined_vars()` gets the variables depending on the scope it's called. https://www.php.net/manual/en/function.get-defined-vars.php

Cheers

Reply:
You can probably use `$relation ?? null` to avoid the phpstan error ?